kiss3/util.py

```diff
@@ -142,3 +146,102 @@ def optional_bool_or_bytes(v):
     if v is not None and not isinstance(v, bool):
         return bytes(v)
     return v
+
+
+_T = TypeVar("_T")
+
+
+@define
+class GenericDecoder(Generic[_T]):
+    """Generic stateful decoder with a callback."""
+
+    callback: Optional[Callable[[_T], None]] = field(default=None)
+
+    @staticmethod
+    def decode_frames(frame: bytes) -> Iterable[_T]:
+        """
+        Decode a single deframed byte chunk.
+
+        :param frame: should represent a single higher level frame to
+            decode in some way.
+        """
+        yield cast(_T, frame)
+
+    def update(self, new_data: bytes) -> Iterable[_T]:
+        """
+        Decode the next sequence of bytes from the stream.
+
+        :param new_data: the next bytes from the stream
+        :return: an iterable of decoded frames
+        """
+        yield cast(_T, new_data)
+
+    def flush(self) -> Iterable[_T]:
+        """Call when the stream is closing to decode any final buffered bytes."""
+        if None:
+            yield
+        return
+
+
+@define
+class FrameDecodeProtocol(asyncio.Protocol, Generic[_T]):
+    """Protocol which uses a GenericDecoder to split the stream into frames."""
+
+    transport: Optional[asyncio.Transport] = field(default=None)
+    decoder: GenericDecoder = field(factory=GenericDecoder)
+    frames: asyncio.Queue = field(factory=asyncio.Queue, init=False)
+    connection_future: asyncio.Future = field(
+        factory=asyncio.Future,
+        init=False,
+    )
+
+    def connection_made(self, transport: asyncio.Transport) -> None:
+        """
+        asyncio callback when connection is established.
+
+        Because this protocol exposes higher-level read/write operations that
+        require the transport, an awaitable `connection_future` is completed for consumers
+        who depend on an active connection.
+        """
+        self.transport = transport
+        self.connection_future.set_result(transport)
+
+    def data_received(self, data: bytes) -> None:
+        """Pass data off to decoder instance and put frames on the queue."""
+        for frame in self.decoder.update(data):
+            self.frames.put_nowait(frame)
+
+    def connection_lost(self, exc: Exception) -> None:
+        """asyncio callback when connection is lost."""
+        for frame in self.decoder.flush():
+            self.frames.put_nowait(frame)
+
+    async def read(self, n_frames=None) -> Iterable[_T]:
+        """
+        Iterate through decoded frames.
+
+        If n_frames is specified, exit after yielding that number of frames.
+        """
+        if n_frames is None:
+            n_frames = -1
+        transport = await self.connection_future
+        while not transport.is_closing() and n_frames:
+            yield await self.frames.get()
+            n_frames -= 1
+        while not self.frames.empty() and n_frames:
+            yield await self.frames.get()
+            n_frames -= 1
```

Comment:
Probably should combine these conditions so the loop continues if either the transport is not closed or the queue is not empty, to avoid subtle edge case